### PR TITLE
feat: casbin access control

### PR DESCRIPTION
**Pull request by Jiraffe7**
Make use of [casbin](https://casbin.org/) for access control.

This implementation consists of
1. Database adapter
2. `CasbinAccessControl`
3. Model configuration `MODEL_CONF` that defines how policies are evaluated. [doc](https://casbin.org/docs/understanding-casbin-detail)

**Database Adapter**
* Loads workspace roles and collab access levels from the database into the model.
* Maps roles to the actions they can perform. E.g. Owner can read/write/delete.

**CasbinAccessControl**
* Provides implementation of existing access control traits.
* Listens for changes to workspace and collab membership to reload policies from the database.

### CODE REVIEW

File: src/config/config.rs [hunk 1]
```diff
@@ -108,6 +114,21 @@ impl DatabaseSetting {
   }
 }
 
+impl ToString for DatabaseSetting {
+  /// Converts [DatabaseSetting] into its postgres connection URL.
+  fn to_string(&self) -> String {
```
Comment (appflowy) [inline src/config/config.rs:119]: Using `to_string` for converting a database connection structure into a PostgreSQL URL might indeed be too vague and not self-explanatory. Creating a dedicated function, such as `to_pg_url`, would be a clearer and more descriptive approach.

File: src/biz/casbin/access_control.rs [hunk 2]
```diff
@@ -0,0 +1,1009 @@
...
+        Ok((obj.to_string(), i32::from(*level).to_string()))
+      },
+      _ => Err(AppError::Internal(anyhow!(
+        "invalid object type and action type combination: object={:?}, action={:?}",
+        obj,
+        act
+      ))),
+    }?;
+
+    let current_policy = self
```
Comment (appflowy) [inline src/biz/casbin/access_control.rs:103]:
```rust
let mut enforcer = self.enforcer.write().await;
    if let Some(policy) = enforcer
        .get_filtered_policy(POLICY_FIELD_INDEX_OBJECT, vec![obj_id.clone()])
        .into_iter()
        .find(|p| p[POLICY_FIELD_INDEX_USER] == uid.to_string()) {
        enforcer
            .remove_policy(policy)
            .await
            .map_err(|e| AppError::Internal(anyhow!("casbin error: {e:?}")))?;
    }

    enforcer
        .add_policy(vec![uid.to_string(), obj_id, action])
        .await
        .map_err(|e| AppError::Internal(anyhow!("casbin error: {e:?}")))
}
```


Holding the write lock of the enforcer throughout the function provides several benefits:

1. Ensures the entire operation (removing an old policy and adding a new one) is atomic. This is crucial for maintaining consistency in access control rules.

2. Prevents race conditions in a multi-threaded environment where multiple threads might try to modify policies simultaneously.

File: src/biz/casbin/access_control.rs [hunk 3]
```diff
@@ -0,0 +1,1009 @@
...
+    let listeners = PgListeners::new(&pool).await?;
+    let access_control = CasbinAccessControl::new(
+      pool.clone(),
+      listeners.subscribe_collab_member_change(),
+      listeners.subscribe_workspace_member_change(),
+      enforcer,
+    );
+    let access_control = access_control.new_workspace_access_control();
+
+    let user = tests::create_user(&pool).await?;
```
Comment (appflowy) [inline src/biz/casbin/access_control.rs:452]: You can find the handy method to create a user in `tests` folder. Try search the: `edit_workspace_without_permission`

File: libs/database/src/collab/collab_db_ops.rs [hunk 4]
```diff
@@ -382,6 +382,26 @@ pub async fn delete_collab_member(uid: i64, oid: &str, pg_pool: &PgPool) -> Resu
   Ok(())
 }
 
+#[inline]
+pub async fn select_all_collab_members(
```
[the remark this answers is not in the document]
Comment (Jiraffe7) [reply libs/database/src/collab/collab_db_ops.rs:386]: Currently this is called on workspace or collab membership changes, when notified by the listeners.

File: src/biz/casbin/access_control.rs [hunk 5]
```diff
@@ -0,0 +1,1009 @@
...
+      .await
+      .enforce((
+        uid.to_string(),
+        ObjectType::Collab(oid).to_string(),
+        action.to_string(),
+      ))
+      .map_err(|e| AppError::Internal(anyhow!("casbin error enforce: {e:?}")))
+  }
+
+  async fn can_send_collab_update(&self, uid: &i64, oid: &str) -> Result<bool, AppError> {
```
[the remark this answers is not in the document]
Comment (Jiraffe7) [reply src/biz/casbin/access_control.rs:333]: Internally, it seems like casbin is implemented using HashMap, so it should be fast enough. If it's not fast enough, we can find other ways to speed it up.

File: src/biz/casbin/access_control.rs [hunk 6]
```diff
@@ -0,0 +1,1009 @@
...
+        "user:{} is not a member of workspace:{}",
+        uid, workspace_id
+      )));
+
+    role
+  }
+}
+
+#[cfg(test)]
+mod tests {
```
[the remark this answers is not in the document]
Comment (Jiraffe7) [reply src/biz/casbin/access_control.rs:412]: The `tests` folder require the server to be running. The intention for these tests is to be lightweight integration tests with only the database.

The existing `tests` folder already exercise the access control from an end-to-end perspective, so one alternative is just to remove the tests in this file if you prefer not to have integration tests here.

File: src/biz/casbin/access_control.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Jiraffe7) [reply src/biz/casbin/access_control.rs:321]: Ok, let me create some benchmarks

**Pull request merged**